src/sentry/search/snuba/executors.py

```diff
@@ -71,134 +40,184 @@ def get_search_filter(search_filters, name, operator):
     return found_val
 
 
+@six.add_metaclass(ABCMeta)
 class AbstractQueryExecutor:
     """This class serves as a template for Query Executors.
     We subclass it in order to implement query methods (we use it to implement two classes: joined Postgres+Snuba queries, and Snuba only queries)
     It's used to keep the query logic out of the actual search backend,
     which can now just build query parameters and use the appropriate query executor to run the query
     """
 
-    def query(
-        self,
-        projects,
-        retention_window_start,
-        group_queryset,
-        environments,
-        sort_by,
-        limit,
-        cursor,
-        count_hits,
-        paginator_options,
-        search_filters,
-        date_from,
-        date_to,
-    ):
-        raise NotImplementedError
+    EMPTY_RESULT = Paginator(Group.objects.none()).get_result()
+    TABLE_ALIAS = ""
 
+    @abstractmethod
+    def query(self, *args, **kwargs):
+        raise NotImplementedError
 
-def snuba_search(
-    start,
-    end,
-    project_ids,
-    environment_ids,
-    sort_field,
-    cursor=None,
-    candidate_ids=None,
-    limit=None,
-    offset=0,
-    get_sample=False,
-    search_filters=None,
-):
-    """
-    This function doesn't strictly benefit from or require being pulled out of the main
-    query method above, but the query method is already large and this function at least
-    extracts most of the Snuba-specific logic.
+    @abstractmethod
+    def calculate_hits(self, *args, **kwargs):
+        raise NotImplementedError
 
-    Returns a tuple of:
-     * a sorted list of (group_id, group_score) tuples sorted descending by score,
-     * the count of total results (rows) available for this query.
-    """
-    filters = {"project_id": project_ids}
+    def _get_dataset(self):
+        if not self.QUERY_DATASET:
+            raise NotImplementedError
+        else:
+            return self.QUERY_DATASET
```

Comment:
This is not what I meant when I suggested to turn this into a method.
What I meant was:

abstract class:
```
@abstractmethod
def _get_dataset(self):
      raise NotImplemented
```

postgres:
```
def _get_dataset(self):
      return snuba.Dataset.Events
```

snuba only
```
def _get_dataset(self):
      return snuba.Dataset.Groups
```

Now you reduced the complexity because the case of `not self.QUERY_DATASET` that you had to handle before and that, in pricnple a callsite would have had to handle before, does not exist anymore.

Some languages allow you to define an abstract constant in an abstract class and the typechecker will ensure you provide a value in all implementation, unfortunately we cannot do that in python, so we have to deal with methods.

Reply:
AbstractQueryExecutor:
```
@property
    @abstractmethod
    def dataset(self):
        """"This function should return an enum from snuba.Dataset (like snuba.Dataset.Events)"""
        raise NotImplementedError
```

Sample implemention:
```
@property
    @abstractmethod
    def dataset(self):
        return snuba.Dataset.Events
```
```